## **Getting Started with Python**

### **1. Hello World Program!**

In [1]:
# printing "Hello World" because its the rule!
print("Hello World!")

Hello World!


**Exploring More!**

In [2]:
print(20+50)

70


In [3]:
print('20+50')

20+50


> Because in `'' or ""` python recognize them as sentence (string). That's why we got as it is value in second time we run the cell `print('20+50')`

In [4]:
print(30>5)

True


In [5]:
print(6<5)

False


> Here, `30` is greater than 5 which is `True` thats why we got True as an output.

### **2. Variables and Data Types**

In [1]:
# creating a python variable is super easy, all we have to do is:
a = 15

**Rules To create a variable in python**

In [2]:
# variable name should not starts with number
1a = 15

SyntaxError: invalid decimal literal (2269455664.py, line 2)

In [3]:
a1 = 15

In [4]:
# variable name should not have space
a 2 = 10

SyntaxError: invalid syntax. Perhaps you forgot a comma? (746718547.py, line 2)

In [5]:
a_2 = 15

**Memory address alocation in python**

In [19]:
a = 10

In [20]:
print("Address of a = ",id(a))

Address of a =  1857617527312


In [21]:
b = a

In [22]:
print("Address of a = ",id(b))

Address of a =  1857617527312


In [23]:
print("Address of b = ",id(b))

Address of b =  1857617527312


In [24]:
print("Is addresses of both variables are same?",id(a)==id(b))

Is addresses of both variables are same? True


In [25]:
# lets change the value of variable a
a = 'Hello'
print(id(a))

1857698929456


**What I did here?**
> As you can see that the new value assigned to a different address which means that the `variable a` still holds the prev data in address = `1857617527312` and new data is stored in new address = `1857698929456`. Which means that this data type cannot be changed therefore we say that this is an **immutable variable**.

### **Mutable & Immutable Data Type**

- **Mutable Data Types (can be modified)**
    - List
    - Dictionary
    - byte array
- **Immutable Data Types (can't be modified)**
    - Int
    - Float
    - Complex
    - String
    - Tuple
    - Set

### **Data Types in Python**
In python we have following data types :
- Numeric
    - Integers 
    - Float
    - Complex Numbers
- Sequence Types
    - String
    - List
    - Tuple
- Dictionary
- Set

#### **Numeric Types**

In [2]:
a = 5
b = 5.5
c = 3.136468436894368684378567897698372689

In [3]:
type(a)

int

In [4]:
type(b)

float

In [5]:
type(c)

float

In [6]:
d = 5j

In [7]:
type(d)

complex

#### **Sequence Data Types**
1. **String Data Types**

In [9]:
varA = 'Hello this is an example of string data type (basically sentences or words)'

In [10]:
# To check type of a variable we use `type()` func
print('Type of varA = ',type(varA))

Type of varA =  <class 'str'>


> **One noteable thing is here that in python we don't have any data types , we have `data classes` instead.**

**Operation on strings**

In [12]:
# string concatination / addition of two or more strings
string1 = 'Hello baby '
string2 = 'How do you do'
print(string1 + string2)

Hello baby How do you do


In [13]:
# another way
print('Hello baby '+'How do you do')

Hello baby How do you do


In [14]:
# capitalization
print(string1.capitalize())

Hello baby 


In [15]:
# uppercase
print(string2.upper())

HOW DO YOU DO


In [16]:
# lower
print(string1.lower())

hello baby 


In [29]:
#  acessing a letter of a string through indexing
string1[0],string1[1],string1[2],string1[3],string1[4],string1[5],string1[6],string1[7],string1[8],string1[9],string1[10],string1[11]

IndexError: string index out of range

> It means in `string1` we only have 12 characters (whitespace included).

In [30]:
string1[0],string1[1],string1[2],string1[3],string1[4],string1[5],string1[6],string1[7],string1[8],string1[9],string1[10]

('H', 'e', 'l', 'l', 'o', ' ', 'b', 'a', 'b', 'y', ' ')

In [43]:
string1[0:9] # slicing string through a start index and end index

'Hello bab'

In [44]:
string1[0:7:2] # slicing strig

'Hlob'

2. **List**

In [45]:
a = [1,2.2,"panda"]

In [46]:
type(a)

list

In [47]:
#  accessing a list value throug idx
a[0]

1

In [48]:
a[1]

2.2

In [50]:
a[2]

'panda'

In [51]:
a[3]

IndexError: list index out of range

In [52]:
# updating a list element
a[0] = 'hello'

In [53]:
a

['hello', 2.2, 'panda']

2. **Tuple**

In [54]:
tuple1 = (1,2,3,0.00000000004,5)

In [55]:
type(tuple1)

tuple

In [60]:
tuple2 = 1,2,3,0.0000000000004,5 # another way of defining a tuple

In [59]:
type(tuple2)

tuple

In [61]:
# accessing a tuple element through idx
tuple1[0]

1

In [62]:
tuple1[2]

3

In [64]:
tuple1[3]

4e-11

In [65]:
# lets try to change a value of tuple
tuple1[0] =10

TypeError: 'tuple' object does not support item assignment

#### **Dictionaries**
A data type of key-value pair

In [66]:
dict1 = {
    "name":['vivek','snehu'],
    "branch":["cse","mech"],
    "class_no":[211,208]
}

In [67]:
dict1["name"]

['vivek', 'snehu']

In [68]:
dict1["branch"]

['cse', 'mech']

In [69]:
dict1["class_no"]

[211, 208]

> good old days :)

In [73]:
# modifying a value in dictionary
dict1["name"] = 'nobody'
dict1["branch"] = 'left the college'
dict1["class_no"] = 'left the college'

In [74]:
dict1

{'name': 'nobody',
 'branch': 'left the college',
 'class_no': 'left the college'}

In [75]:
# adding a new-key value pair
dict1["what_happend"] = 'end'
dict1['betrayal'] = True

In [76]:
dict1

{'name': 'nobody',
 'branch': 'left the college',
 'class_no': 'left the college',
 'what_happend': 'end',
 'betrayal': True}

> sad ending :(

#### **Set**

In [77]:
set1 = {'anshu','siru','dhanashree'}

In [80]:
type(set1)

set

In [78]:
set2 =  {'anshu','siru','dhanashree','anshu'}

In [79]:
set2

{'anshu', 'dhanashree', 'siru'}

> set removes the duplicates just as I do :)